category_theory/isomorphism.lean

```diff
@@ -0,0 +1,160 @@
+      end,
+    induction p, induction w,
+    refl
+  end
+
+@[refl] def refl (X : C) : X ≅ X := 
+{ hom := 𝟙 X,
+  inv := 𝟙 X }
+
+@[simp] lemma refl_map (X : C) : (iso.refl X).hom = 𝟙 X := rfl
+@[simp] lemma refl_coe (X : C) : ((iso.refl X) : X ⟶ X) = 𝟙 X := rfl
+@[simp] lemma refl_inv  (X : C) : (iso.refl X).inv  = 𝟙 X := rfl
+
+@[trans] def trans (α : X ≅ Y) (β : Y ≅ Z) : X ≅ Z := 
+{ hom := α.hom ≫ β.hom,
+  inv := β.inv ≫ α.inv,
+  hom_inv_id' := begin /- `obviously'` says: -/ erw [category.assoc], conv { to_lhs, congr, skip, rw ← category.assoc }, rw iso.hom_inv_id, rw category.id_comp, rw iso.hom_inv_id end,
+  inv_hom_id' := begin /- `obviously'` says: -/ erw [category.assoc], conv { to_lhs, congr, skip, rw ← category.assoc }, rw iso.inv_hom_id, rw category.id_comp, rw iso.inv_hom_id end }
+
+infixr ` ≪≫ `:80 := iso.trans -- type as `\ll \gg`.
+
+@[simp] lemma trans_hom (α : X ≅ Y) (β : Y ≅ Z) : (α ≪≫ β).hom = α.hom ≫ β.hom := rfl
+@[simp] lemma trans_coe (α : X ≅ Y) (β : Y ≅ Z) : ((α ≪≫ β) : X ⟶ Z) = (α : X ⟶ Y) ≫ (β : Y ⟶ Z) := rfl
+@[simp] lemma trans_inv (α : X ≅ Y) (β : Y ≅ Z) : (α ≪≫ β).inv  = β.inv ≫ α.inv   := rfl
+
+@[symm] def symm (I : X ≅ Y) : Y ≅ X := 
+{ hom := I.inv,
+  inv := I.hom }
+
+@[simp] lemma refl_symm_coe (X : C) : ((iso.refl X).symm : X ⟶ X)  = 𝟙 X := rfl
+@[simp] lemma trans_symm_coe (α : X ≅ Y) (β : Y ≅ Z) : ((α ≪≫ β).symm : Z ⟶ X) = (β.symm : Z ⟶ Y) ≫ (α.symm : Y ⟶ X) := rfl
+
+-- These next two aggressively rewrite the projections `.hom` and `.inv` into coercions.
+-- I'm not certain it is a good idea.
+@[simp] lemma hom_coe (α : X ≅ Y) : α.hom = (α : X ⟶ Y) := rfl
+@[simp] lemma inv_coe (α : X ≅ Y) : α.inv = (α.symm : Y ⟶ X) := rfl
+
+-- FIXME these are actually the ones we want to use
+@[simp] lemma hom_inv_id_coe (α : X ≅ Y) : (α : X ⟶ Y) ≫ (α.symm : Y ⟶ X) = 𝟙 X := begin unfold_coes, unfold symm, rw iso.hom_inv_id, end
+@[simp] lemma inv_hom_id_coe (α : X ≅ Y) : (α.symm : Y ⟶ X) ≫ (α : X ⟶ Y) = 𝟙 Y := begin unfold_coes, unfold symm, rw iso.inv_hom_id, end
+
+end iso
+
+class is_iso (f : X ⟶ Y) :=
+(inv : Y ⟶ X)
+(hom_inv_id' : f ≫ inv = 𝟙 X . obviously)
+(inv_hom_id' : inv ≫ f = 𝟙 Y . obviously)
+
+restate_axiom is_iso.hom_inv_id'
+restate_axiom is_iso.inv_hom_id'
+attribute [simp,ematch] is_iso.hom_inv_id is_iso.inv_hom_id
+
+def inv' {f : X ⟶ Y} (p : is_iso f) := is_iso.inv f 
+def hom_inv_id' {f : X ⟶ Y} (p : is_iso f) : f ≫ inv' p = 𝟙 X := is_iso.hom_inv_id f 
+def inv_hom_id' {f : X ⟶ Y} (p : is_iso f) : inv' p ≫ f = 𝟙 Y := is_iso.inv_hom_id f 
+
+namespace is_iso
+
+instance (X : C) : is_iso (𝟙 X) := 
+{ inv := 𝟙 X }
+
+instance of_iso         (f : X ≅ Y) : is_iso (f : X ⟶ Y) :=
+{ inv   := f.inv }
+instance of_iso_inverse (f : X ≅ Y) : is_iso (f.symm : Y ⟶ X)  := 
+{ inv   := f.hom }
+
+end is_iso
+
+namespace functor
+
+universes u₁ v₁ u₂ v₂ 
+variables {D : Type u₂}
+
+variables [𝒟 : category.{u₂ v₂} D]
+include 𝒟
+
+def on_isos (F : C ↝ D) {X Y : C} (i : X ≅ Y) : (F X) ≅ (F Y) :=
```

Comment:
`on_iso`? it works only on one isomorphism.

Reply:
Fixed, thanks.